components/higher-order/with-safe-timeout/index.js

```diff
@@ -0,0 +1,56 @@
+/**
+ * WordPress dependencies
+ */
+import { Component } from '@wordpress/element';
+
+/**
+ * Browser dependencies
+ */
+const { clearTimeout, setTimeout } = window;
+
+/**
+ * A higher-order component used to provide and manage delayed function calls
+ * that ought to be bound to a component's lifecycle.
+ *
+ * @param {Component} OriginalComponent   Component requiring setTimeout
+ *
+ * @returns {Component}                   Wrapped component.
+ */
+function withSafeTimeout( OriginalComponent ) {
+	return class WrappedComponent extends Component {
+		constructor() {
+			super( ...arguments );
+			this.timeouts = [];
+			this.setTimeout = this.setTimeout.bind( this );
+			this.clear = this.clear.bind( this );
+		}
+
+		componentWillUnmount() {
+			this.timeouts.forEach( clearTimeout );
+		}
+
+		setTimeout( fn, delay ) {
+			const id = setTimeout( () => {
+				fn();
+				this.clear( id );
+			}, delay );
+			this.timeouts.push( id );
+		}
+
+		clear( id ) {
+			const newTimeouts = this.timeouts.filter( t => t !== id );
+			this.timeouts = newTimeouts;
+		}
+
+		render() {
+			return (
+				<OriginalComponent
+					{ ...this.props }
+					setTimeout={ this.setTimeout }
```

Comment:
Should we return the `timeoutID`, in case someone wants to clear the timeout not just on unmount? Should we pass `clearTimeout` as well?

Reply:
> Should we return the `timeoutID`

Absolutely, which I just noticed I didn't do.

> Pass `clearTimeout`

This I'm not so sure. What would we provide? The same as `window.clearTimeout`?

Reply:
Yes. It might be more convenient to get it from the same place. Without it I'd need to get one from props, and "import" the other from the window. Maybe it's also more future-proof.

Reply:
Also, I guess to be super correct, it would also need to call `this.clear( id )`.